libraries/standard/mqtt/src/mqtt.c

```diff
@@ -504,10 +1009,90 @@ MQTTStatus_t MQTT_Disconnect( MQTTContext_t * const pContext )
 
 /*-----------------------------------------------------------*/
 
-MQTTStatus_t MQTT_Process( MQTTContext_t * const pContext,
-                           uint32_t timeoutMs )
+MQTTStatus_t MQTT_ProcessLoop( MQTTContext_t * const pContext,
+                               uint32_t timeoutMs )
 {
-    return MQTTSuccess;
+    MQTTStatus_t status = MQTTBadParameter;
+    MQTTGetCurrentTimeFunc_t getTimeStamp = NULL;
+    uint32_t entryTime = 0, remainingTimeMs = timeoutMs;
+    MQTTPacketInfo_t incomingPacket;
+
+    if( pContext != NULL )
+    {
+        getTimeStamp = pContext->callbacks.getTime;
+        entryTime = getTimeStamp();
+        status = MQTTSuccess;
+    }
+    else
+    {
+        IotLogError( "MQTT Context cannot be NULL." );
+    }
+    
+    while( ( pContext != NULL ) && 
+           ( calculateElapsedTime( getTimeStamp(), entryTime ) < timeoutMs ) )
```

Comment:
What is the expected behavior when 0 is passed as timeoutMs?

Reply:
This should execute at least once even with 0 timeout, similar to a do-while. I'll move this check to inside the loop